### DISCUSSION TITLE
Is There A Way To Read/Edit The Current DL Listing?

### DISCUSSION THREAD

**Discussion by DeathStalker77** [nicotine-plus/nicotine-plus#2687] [Q&A]
For v3.3.0 B5 from 10/09?

I need to be able to see the DLs as they would appear in the normal DL window, so I can edit/delete stuff, as I currently cannot do **ANYTHING** with the program - it just hangs.

Thanks.

--- DS

**Comment by slook** [comment]
Not really no. It would be ideal if line-breaks were inserted into the `downloads.json` and `upload.json` files, similar to how the shareslist files are now formatted. However, you could try using a web browser or other xml viewer to see the json files more clearly.

To avoid the hang you can move, rename or delete the json files (while the application is not running).

**Comment by DeathStalker77** [comment]
Yes, that was one of my thoughts as well.  Currently, I see no way to effectively parse it out because of all the special characters that are not only used as "standard", but that can also be used as part of the file/path name.  There _**HAS**_ to be something unique used to separate individual USERS, and also the individual files of each USER.  To me, this is only "common sense".

If I rename the files, then I LOSE everything I currently have!  I do NOT want to do that.  It's a pretty simple request, impo.  I HAVE to be able to separate things - even if it is just by USER.  Not just LOSE them.

I can open the JSON in WORD, but again, the formatting of the file is JUST SO BAD, trying to make sense of it would basically take hours (or the use of GenAI).

**Reply by slook** [reply]
> There HAS to be something unique

Ideally, it would be either a carriage return, or a line break.

Each transfer object ends with `}],`

**Comment by DeathStalker77** [comment]
Also, having just searched through the download.json file, I find that there are users that I have unfinished files from in my folder, that are NOT in the file ..... strange.

it **seems** that the start of a USER begins with [ but ends with just a comma & space, before the next User's [.

VERY strange "formatting", impo.

Might I suggest a conversion into a "proper" database?  With proper columns, etc?  MSMSQL or even SQLite or MySQL.  With that, not only would the data be more secure and less prone to corruption, but also easily editable.  (Will add this to the Feature Request).  Even qBitTorrent has switched to a SQLite DB.

--- DS

**Reply by slook** [reply]
It is a list of lists in standard JSON format, structured as follows:

```
[
  [
    transfer.username,
    transfer.virtual_path,
    transfer.folder_path,
    transfer.status,
    transfer.size,
    transfer.current_byte_offset,
    transfer.file_attributes {dict}
  ],
  [
    next transfer
  ],
  [
    ...
  ],
  [
    last transfer
  ]
]
```

- Each `transfer` object is contained in `[`square brackets`]`
- The last parameter `file_attributes` is a dict, so it is surrounded with `{`curly brackets`}`
- The last list item in each list doesn't need a comma `,` at the end]
- Each transfer object ends with `"}],`
- Whitespace (spaces, tabs, newlines) are optional and ignored.

**Comment by DeathStalker77** [comment]
Ok, but with respect, that still does not provide a way to effectively re-format the file.  If it is a "standard JSON format", opening it in Notepad++ should display it properly - it doesn't - it displays it as one continuous line.

**Reply by slook** [reply]
It looks like the line breaks are replaced with the underscore `_` replacement character, which indeed is useless as you have respectfully highlighted.

**Reply by DeathStalker77** [reply]
Actually, I only found 1 underscore in the whole file, so that can't be the line break character.

**Reply by slook** [reply]
You're correct, I was mistaken about that. My eyes just kinda go all blury when I look at an unbroken wall of gibberish like that.

**Comment by DeathStalker77** [comment]
Well, it would seem as if there is some development to be done there.  Again, might I suggest using a proper database with specific columns?  That would help tremendously (and probably improve performance.  JSONs have an "effective" limit with the amount of data they can carry.  I encountered this when I was tasked with working on a DB project that wanted to use only JSONs for input & output (that also included al DQ aspects) - as a DQ expert, I advised that it could not be done.  In the end, the devs had to admit that I was right & the program had to be canceled.  Now I'm not saying that your current JSON has the DQ issues, but it DOES have the issue of being tasked with trying to convey too much data.  I'm not even certain that a separate JSON per DL USERNAME would be effective.

**Comment by slook** [comment]
I just discovered that the LibreWolf web browser (and FireFox or any Gecko based browser I presume) has a JSON renderer in it, which enables the expanded viewing and Pretty Print of downloads.json and uploads.json that can be saved (though sadly, partial copy/export is not possible) then you can open that in Notepad++ for manual editing.

**Reply by DeathStalker77** [reply]
Interesting!  It's off than NONE of the other JSON parsers (whether online or not) or browsers, will display it in that manner!  Is "Pretty Print" a separate program to be installed, or is it part of the browser/s?

**UPDATE:  Actually, it seems even EDGE can open it properly formatted - I never even thought of using a browser! :-)

**Reply by slook** [reply]
It's built right in, just Open with... https://librewolf.net/

![image](https://github.com/nicotine-plus/nicotine-plus/assets/88614182/8e549109-a9df-4f86-abac-4510e45ac235)

**Reply by DeathStalker77** [reply]
Installing Librewolf - interestingly, while EDGE opened the file with no problem, copy/pasting it to WORD appears to have exceeded WORD's limitations - I have never encountered that before!

**Comment by mathiascode** [comment]
We used to have line breaks in the json files, but had to disable them since Python's faster, C-based json encoder didn't support them. I'll see if we can work around this somehow.

As for moving to SQLite or similar, I'm not opposed to the idea, but it won't happen right now. I'll have to see what the best option is moving forward when I work on the transfer-related modules again.

**Reply by DeathStalker77** [reply]
Ok, thanks!  I am happy that you are at least looking into the possibility of some type of proper DB :-)  Again, I understand Priorities vs Enhancements :-)

**Reply by mathiascode** [reply]
The latest development build shows one transfer per line in the json files. Keep in mind that you need to exit Nicotine+ before editing the file manually.

**Reply by DeathStalker77** [reply]
Ok, Yes, you can OPEN the JSON file in LibreWolf, but you cannot EDIT in it.

So what program can _effectively_ handle the editing?

**Reply by mathiascode** [reply]
Any text editor that's able to display a large number of lines should do, e.g. Notepad (at least it used to be good at this). What changes are you trying to make to the list?

**Reply by DeathStalker77** [reply]
Now that you have changed the lines to a single record per line, it will open & edit in Notepad++  (IMPO, I still think there should be some sort of LF/CR included)

--- DS
